### PR TITLE
CurrencyAvgRate table data to calculate p&l amount

### PR DESCRIPTION
**Pull request by Himanshunegi0332**
Targets #2487 

### Description
Created a CurrencyAvgRate table data to calculate p&l amount If there is no entry for that month then we can use the current rates by default.

### Checklist:
- [X] I have performed a self-review of my own code.

### CODE REVIEW

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 1]
```diff
@@ -59,12 +60,19 @@ private function getParticularAmountForExport(array $particular, Object $startDa
...
-        $exchangeRates = app(CurrencyService::class)->getCurrentRatesInINR();
-
+        $exchangeRates = CurrencyAvgRate::select('avg_rate', 'captured_for')->groupBy('captured_for')->get()->toArray();
+        foreach ($exchangeRates as $exchangeRate) {
+            if ($exchangeRate['avg_rate']) {
+                $exchangeDollor = $exchangeRate['avg_rate'];
+            } else {
+                $exchangeDollor = app(CurrencyService::class)->getCurrentRatesInINR();
+            }
+        }
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:73]: We don't need this loop or something. 

We just need a value where 
captured_for month is the same as invoice->sent->date month, 
currency  = invoice->currency

**Review by pankaj-ag — CHANGES_REQUESTED** [review]

File: Modules/Report/Services/Finance/RevenueReportService.php (hunk 1 above)
Comment (Himanshunegi0332) [reply Modules/Report/Services/Finance/RevenueReportService.php:73]: > We don't need this loop or something.
> 
> We just need a value where captured_for month is the same as invoice->sent->date month, currency = invoice->currency

Can you explain it little bit.

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 2]
```diff
@@ -80,49 +88,26 @@ private function getParticularAmountForExport(array $particular, Object $startDa
 
     private function getParticularAmountForCommissionReceived(array $particular, Object $startDate, Object $endDate): array
     {
-        return $this->getAmountsForRevenueProceeds(Str::snake($particular['name']), $startDate, $endDate);
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:83]: @Himanshunegi0332 why w reverted those changes ?

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 3]
```diff
@@ -80,49 +88,26 @@ private function getParticularAmountForExport(array $particular, Object $startDa
...
-            $month = sprintf('%02d', $revenue->month);
-            $dateKey = $month . '-' . $year;
-            $totalAmount += $amount;
-            $results[$dateKey] = ($results[$dateKey] ?? 0) + $amount;
-        }
-        $results['total'] = $totalAmount;
-
-        return $results;
+        return ['total' => 0];
     }
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:112]: Why we removed this code? Please add it back.

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 4]
```diff
@@ -59,13 +59,21 @@ private function getParticularAmountForExport(array $particular, Object $startDa
...
+        $exchangeDollor = 0;
 
         // ToDo:: We need to change this logic and get the exchange rate for every month.
-        $exchangeRates = app(CurrencyService::class)->getCurrentRatesInINR();
-
+        $exchangeRates = CurrencyAvgRate::select('avg_rate', 'captured_for')->groupBy('captured_for')->get()->toArray();
         foreach ($invoices as $invoice) {
-            $amount = $invoice->amount * $exchangeRates;
             $dateKey = $invoice->sent_on->format('m-y');
+            foreach ($exchangeRates as $exchangeRate) {
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:68]: Let's not run this loop again and again for each invoice. 
Let's take this loop outside of the top for loop and store the values in a variable like currency_chages [
'01-22' => 74,
'02-22' => 0,
 
]

and then use that variable.

**Review by pankaj-ag — CHANGES_REQUESTED** [review]

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 5]
```diff
@@ -59,17 +60,29 @@ private function getParticularAmountForExport(array $particular, Object $startDa
...
+            $exchangeMonth = (date('m-y', strtotime($exchangeRate['captured_for'])));
+            $currency_chages[] = [
+                $exchangeMonth => $exchangeRate['avg_rate']
+            ];
+        }
         foreach ($invoices as $invoice) {
-            $amount = $invoice->amount * $exchangeRates;
             $dateKey = $invoice->sent_on->format('m-y');
+            if ($exchangeMonth == $dateKey) {
+                $exchangeDollor = $exchangeRate['avg_rate'];
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:78]: I think this should be like this right ?

```suggestion
            if (isset($currency_chages[$dateKey])) {
                $exchangeDollor = $currency_chages[$dateKey];
```

Comment (pankaj-ag) [reply Modules/Report/Services/Finance/RevenueReportService.php:78]: Please test by checking if the invoice amount is calculating based on the other the currency value. 

Like if a April invoice amount is 100
and April avg rate is 70 

Then the invoice value will be 7000

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 6]
```diff
@@ -59,17 +60,29 @@ private function getParticularAmountForExport(array $particular, Object $startDa
...
-
+        $exchangeRate = [];
+        $exchangeRates = CurrencyAvgRate::select('avg_rate', 'captured_for')->groupBy('captured_for')->get()->toArray();
+        $exchangeMonth = null;
+        foreach ($exchangeRates as $exchangeRate) {
+            $exchangeMonth = (date('m-y', strtotime($exchangeRate['captured_for'])));
+            $currency_chages[] = [
+                $exchangeMonth => $exchangeRate['avg_rate']
+            ];
+        }
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:74]: Very good now let's move this to new function with name like getAvgCurrencyRates()

**Review by pankaj-ag — CHANGES_REQUESTED** [review]

File: Modules/Report/Services/Finance/RevenueReportService.php (hunk 5 above)
Comment (Himanshunegi0332) [reply Modules/Report/Services/Finance/RevenueReportService.php:78]: $exchangeDollor = $currency_chages[$dateKey];     by doing this we will get only the month and year not the avg_rate

File: Modules/Report/Services/Finance/RevenueReportService.php (hunk 6 above)
Comment (Himanshunegi0332) [reply Modules/Report/Services/Finance/RevenueReportService.php]: ok done.

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 7]
```diff
@@ -37,9 +38,27 @@ public function getParticularReport(String $particularSlug, array $particular, i
...
+        $totalRate = $exchangeRate['avg_rate'];
+
+        return $totalRate;
+    }
+
     private function getParticularAmountForDomestic(array $particular, Object $startDate, Object $endDate): array
     {
-        $invoices = $this->invoiceService->getInvoicesBetweenDates($startDate, $endDate, 'indian');
+        $invoices = $this->invoiceService->getInvoicesBetweenDates($startDate, $endDate, 'non-indian');
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:61]:
```suggestion
        $invoices = $this->invoiceService->getInvoicesBetweenDates($startDate, $endDate, 'indian');
```

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 8]
```diff
@@ -59,17 +78,24 @@ private function getParticularAmountForExport(array $particular, Object $startDa
...
+        }
 
         // ToDo:: We need to change this logic and get the exchange rate for every month.
-        $exchangeRates = app(CurrencyService::class)->getCurrentRatesInINR();
-
         foreach ($invoices as $invoice) {
-            $amount = $invoice->amount * $exchangeRates;
             $dateKey = $invoice->sent_on->format('m-y');
+            if ($exchangeMonth == $dateKey) {
+                $exchangeDollor = $this->getAvgCurrencyRates();
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:91]: @Himanshunegi0332 please try to understand the code first and then write :).

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 9]
```diff
@@ -37,9 +38,27 @@ public function getParticularReport(String $particularSlug, array $particular, i
         return $particular;
     }
 
+    public function getAvgCurrencyRates()
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:41]: OKay now let's send the startDate and endDate in this function and get the currency average between those dates only.

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 10]
```diff
@@ -59,17 +78,24 @@ private function getParticularAmountForExport(array $particular, Object $startDa
         $invoices = $this->invoiceService->getInvoicesBetweenDates($startDate, $endDate, 'non-indian');
         $totalAmount = 0;
         $results = [];
+        $exchangeMonth = null;
+        $exchangeRates = CurrencyAvgRate::select('captured_for')->groupBy('captured_for')->get()->toArray();
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:82]:
```suggestion
        $exchangeRates = $this->getAvgCurrencyRates($startDate, $endDate);
```

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 11]
```diff
@@ -59,17 +78,24 @@ private function getParticularAmountForExport(array $particular, Object $startDa
...
+        foreach ($exchangeRates as $exchangeRate) {
+            $exchangeMonth = (date('m-y', strtotime($exchangeRate['captured_for'])));
+        }
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:85]:
```suggestion
       
```

File: Modules/Report/Services/Finance/RevenueReportService.php [hunk 12]
```diff
@@ -59,17 +78,24 @@ private function getParticularAmountForExport(array $particular, Object $startDa
...
 
         // ToDo:: We need to change this logic and get the exchange rate for every month.
-        $exchangeRates = app(CurrencyService::class)->getCurrentRatesInINR();
-
         foreach ($invoices as $invoice) {
-            $amount = $invoice->amount * $exchangeRates;
             $dateKey = $invoice->sent_on->format('m-y');
+            if ($exchangeMonth == $dateKey) {
```
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:90]:
```suggestion
            if (isset($exchangeRates[$dateKey])) { 
```

File: Modules/Report/Services/Finance/RevenueReportService.php (hunk 8 above)
Comment (pankaj-ag) [inline Modules/Report/Services/Finance/RevenueReportService.php:91]:
```suggestion
                $exchangeDollor = $exchangeRates[$dateKey];
```

**Review by pankaj-ag — CHANGES_REQUESTED** [review]

File: Modules/Report/Services/Finance/RevenueReportService.php (hunk 5 above)
Comment (pankaj-ag) [reply Modules/Report/Services/Finance/RevenueReportService.php]: So please why this is not working. 

Also check the exchangeRates array format. It should be
[
"month-Year" => "AvgRates"
]

File: Modules/Report/Services/Finance/RevenueReportService.php (hunk 7 above)
Comment (pankaj-ag) [reply Modules/Report/Services/Finance/RevenueReportService.php:61]: Sorry it was my bad.

File: Modules/Report/Services/Finance/RevenueReportService.php (hunk 10 above)
Comment (Himanshunegi0332) [reply Modules/Report/Services/Finance/RevenueReportService.php:82]: ok sir

**Review by pankaj-ag — APPROVED** [review]

**Review by pankaj-ag — APPROVED** [review]

**Pull request merged**